common/src/block_interval.rs

```diff
@@ -0,0 +1,255 @@
+    ///    assert_eq!(BlockInterval::new("100..").unwrap(), BlockInterval::FollowFrom(100));
+    /// ```
+    pub fn new(s: &str) -> Result<BlockInterval, BlockIntervalError> {
+        if (s.starts_with("0x") && s.len() == 66) || s.len() == 64 {
+            // Try to parse hash
+            let hash = s
+                .parse::<B256>()
+                .map_err(|_| BlockIntervalError::InvalidBlockHash(s.to_string()))?;
+            return Ok(BlockInterval::SingleBlockId(BlockId::Hash(hash.into())));
+        }
+
+        // First we parse for inclusive range and then for exclusive range,
+        // because both separators start with `..`
+        if let Ok(range) = parsing::parse_range_inclusive(s) {
+            Ok(BlockInterval::Range(range))
+        } else if let Ok(range) = parsing::parse_range_exclusive(s) {
+            Ok(BlockInterval::Range(range))
+        }
+        // Now we look for the follow from range
+        else if s.contains("..") {
+            let mut split = s.trim().split("..").filter(|s| *s != "=" && !s.is_empty());
+
+            // Any other character after `..` or `..=` is invalid
+            if split.clone().count() > 1 {
+                return Err(BlockIntervalError::InvalidFollowFromRange(s.to_string()));
+            }
+            let num = split
+                .next()
+                .map(|num| {
+                    num.parse::<u64>()
+                        .map_err(|_| BlockIntervalError::InvalidBlockNumber(s.to_string()))
+                })
+                .ok_or(BlockIntervalError::InvalidBlockIntervalRange(s.to_string()))??;
+            return Ok(BlockInterval::FollowFrom(num));
+        }
+        // Only single block number is left to try to parse
+        else {
+            let num = s
+                .trim()
+                .parse()
+                .map_err(|_| BlockIntervalError::InvalidBlockIntervalRange(s.to_string()))?;
+            return Ok(BlockInterval::Single(num));
+        }
+    }
+
+    /// Convert the block interval into an async stream of block numbers.
+    pub fn into_bounded_stream(self) -> Result<impl Stream<Item = u64>, BlockIntervalError> {
+        match self {
+            BlockInterval::Single(num) => Ok(tokio_stream::iter(num..num + 1)),
+            BlockInterval::Range(range) => Ok(tokio_stream::iter(range)),
+            _ => Err(BlockIntervalError::IntoBoundedStreamError),
+        }
+    }
+
+    /// Convert the block interval into an unbounded async stream of block
+    /// numbers. Query the blockchain node for the latest block number.
+    pub async fn into_unbounded_stream<ProviderT, TransportT>(
+        self,
+        provider: ProviderT,
+    ) -> Result<impl Stream<Item = Result<u64, anyhow::Error>>, BlockIntervalError>
+    where
+        ProviderT: Provider<TransportT>,
+        TransportT: Transport + Clone,
+    {
+        match self {
+            BlockInterval::FollowFrom(start) => Ok(try_stream! {
```

Comment:
I think you can write this without a macro see [`try_unfold`](https://docs.rs/futures/latest/futures/stream/fn.try_unfold.html)

Reply:
Yes, it could be probably rewritten with `try_unfold`. I will probably work more on `follow-from` scenario later (we need some integration tests for it) then I could try to optimize it.